docs/howtos/testing-a-rust-component.md

```diff
@@ -0,0 +1,146 @@
+[XCode unittest framework](https://developer.apple.com/documentation/xctest).
+
+Due to the way that all rust components need to be compiled together into a single ["megazord"](../design/megazords.md)
+framework, this entire respository is a single XCode project. The Swift tests for each component
+thus need to live under `megazords/ios/MozillaAppServicesTests/` rather than in the directory
+for the corresponding component. (XXX TODO: is this true? it would be nice to find a way to avoid havining
+them live separately because it makes them easy to overlook).
+
+The Swift tests can be `./gradlew <component>:test`.
+
+The tests at this layer are designed to ensure that the API binding code is working as intended,
+and should not repeat tests for functionality that is already well tested at the Rust level.
+But given that the Swift bindings involve a non-trivial amount of hand-written boilerplate code,
+it's important to exercise that code throughly.
+
+(Editor's note: I remain hopeful that one day we'll autogenerate most of the Swift binding code, and in
+such a world we don't need to invest in tests for it.)
+
+XXX TODO: any additional tips here, such as mocking out storage etc?
+
+## Integration tests
+
+### End-to-end Sync Tests
+
+The [`testing/sync-test`](../../testing/sync-test) directory contains a test harness for running sync-related
+Rust components against a live Firefox Sync infrastructure, so that we can verifying the functionality
+end-to-end.
+
+Each component that implements a sync engine should have a corresponding suite of tests in this directory.
+
+* XXX TODO: places doesn't.
+* XXX TODO: send-tab doesn't (not technically a sync engine, but still, it's related)
+* XXX TODO: sync-manager doesn't
+
+### Android Components Test Suite
+
+It's important that changes in application-services are tested against upstream consumer code in the
+[android-components](https://github.com/mozilla-mobile/android-components/) repo. This is currently
+a manual process involving:
+
+* Configuring your local checkout of android-components to [use your local application-services
+  build](./working-with-reference-browser.md).
+* Running the android-components test suite via `./gradle test`.
+* Manually building and running the android-components sample apps to verify that they're still working.
+
+Ideally some or all of this would be automated and run in CI, but we have not yet invested in such automation.
+
+## Test Coverage
+
+Lamentably, we do not measure or report on code test coverage.
+See [this github issue](https://github.com/mozilla/application-services/issues/1745) for some early explorations.
+
+The rust ecosystem for code coverage is still maturing, with [cargo-tarpaulin](https://github.com/xd009642/tarpaulin)
+appearing to be a promising candidate.  However, such tools will only report code that is exercised by the rust
+unittests, not code that is exercised by the Kotlin or Switch tests or the end-to-end integration test suite.
```

Comment:
Nit: Swift

Reply:
LOL I may have written this in the lead-up the Christmas, wherein I was excited to be getting a certain gaming console for the family...